controllers/podchaos/containerkill/types.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containerkill
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/go-logr/logr"
+
+	"github.com/pingcap/chaos-mesh/api/v1alpha1"
+	pb "github.com/pingcap/chaos-mesh/pkg/chaosdaemon/pb"
+	"github.com/pingcap/chaos-mesh/pkg/utils"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	containerKillActionMsg = "delete container %s"
+)
+
+type Reconciler struct {
+	client.Client
+	Log logr.Logger
+}
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var err error
+	now := time.Now()
+
+	r.Log.Info("reconciling container kill")
+	ctx := context.Background()
+
+	var podchaos v1alpha1.PodChaos
+	if err = r.Get(ctx, req.NamespacedName, &podchaos); err != nil {
+		r.Log.Error(err, "unable to get podchaos")
+		return ctrl.Result{}, nil
+	}
+
+	if podchaos.Spec.ContainerName == "" {
+		r.Log.Error(nil, "the name of container is empty", "name", req.Name, "namespace", req.Namespace)
+		return ctrl.Result{}, nil
+	}
+
+	shouldAct := podchaos.GetNextStart().Before(now)
+	if !shouldAct {
+		return ctrl.Result{RequeueAfter: podchaos.GetNextStart().Sub(now)}, nil
+	}
+	pods, err := utils.SelectPods(ctx, r.Client, podchaos.Spec.Selector)
+	if err != nil {
+		r.Log.Error(err, "fail to get selected pods")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if len(pods) == 0 {
+		r.Log.Error(nil, "no pod is selected", "name", req.Name, "namespace", req.Namespace)
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	filteredPod, err := utils.GeneratePods(pods, podchaos.Spec.Mode, podchaos.Spec.Value)
+	if err != nil {
+		r.Log.Error(err, "fail to generate pods")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	g := errgroup.Group{}
+	haveSelected := false
+	for podIndex := range filteredPod {
+		pod := &filteredPod[podIndex]
+
+		for containerIndex := range pod.Status.ContainerStatuses {
+			containerName := pod.Status.ContainerStatuses[containerIndex].Name
+			containerID := pod.Status.ContainerStatuses[containerIndex].ContainerID
+
+			if containerName == podchaos.Spec.ContainerName {
+				haveSelected = true
```

Comment:
I think `haveSelected` is unnecessary, because If the only one from all selected pod met this name, we will think all pod meets the name.